yt_idv/shaders/ray_tracing.frag.glsl

```diff
@@ -45,6 +50,106 @@ vec4 cleanup_phase(in vec4 curr_color, in vec3 dir, in float t0, in float t1);
 //   void (vec3 tex_curr_pos, inout vec4 curr_color, float tdelta, float t,
 //         vec3 direction);
 
+float get_ray_plane_intersection(vec3 p_normal, float p_constant, vec3 ray_origin, vec3 ray_dir)
+{
+    float n_dot_u = dot(p_normal, ray_dir);
+    float n_dot_ro = dot(p_normal, ray_origin);
+    // add check for n_dot_u == 0 (ray is parallel to plane)
+    if (n_dot_u == float(0.0))
+    {
+        // the ray is parallel to the plane. there are either no intersections
+        // or infinite intersections. In the second case, we are guaranteed
+        // to intersect one of the other faces, so we can drop this plane.
+        return INFINITY;
+    }
+
+    return (p_constant - n_dot_ro) / n_dot_u;
+}
+
+vec2 get_ray_sphere_intersection(float r, vec3 ray_origin, vec3 ray_dir)
+{
+    float dir_dot_dir = dot(ray_dir, ray_dir);
+    float ro_dot_ro = dot(ray_origin, ray_origin);
+    float dir_dot_ro = dot(ray_dir, ray_origin);
+    float rsq = r * r; // could be calculated in vertex shader
```

Comment:
which r is this?  could we actually calculate it in the vertex shader?  eventually this would need to be once for every *cell* edge, right?

Reply:
that would be the r from `left_edge` or `right_edge`. So I could calculate r^2 for both of those in the vertex shader, ya?